include/revng/AutoEnforcer/AutoEnforcer.h

```diff
@@ -0,0 +1,183 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <utility>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/AutoEnforcer/AutoEnforcerContext.h"
+#include "revng/AutoEnforcer/AutoEnforcerTarget.h"
+#include "revng/AutoEnforcer/BackingContainerRegistry.h"
+#include "revng/AutoEnforcer/InvalidationEvent.h"
+#include "revng/AutoEnforcer/KindsRegistry.h"
+#include "revng/AutoEnforcer/Pipeline.h"
+#include "revng/Support/Debug.h"
+
+namespace AutoEnforcer {
+
+class PipelineRunner {
+public:
+  using iterator = std::vector<Step>::iterator;
+  using const_iterator = std::vector<Step>::const_iterator;
+  using InvalidationMap = llvm::StringMap<BackingContainersStatus>;
+
+  explicit PipelineRunner(Context &C) : AEContext(&C) {}
+
+  template<typename... EnforcerWrappers>
+  void addStep(std::string StepName, EnforcerWrappers &&...EnfWrappers) {
```

Comment:
* [ ] `Enf...`, no abbreviations.
* [ ] I'm a bit confused. If this is the only way in which a `PipelineRunner` can be built, and it's strictly compile-time, how can we build from YAML?

Reply:
the pipeline loader invokes the correct overload by keeping a table of names to way of wrapping up enforcers.

Reply:
Can you show it to me?

Reply:
ah no i got confused, the loader uses the specialization with no arguments and then it manually adds the various steps one at the time starting from the step reference.

the version with more than one argument is used in the tests and in revng-lift
```
AE.addStep("first_step", bindEnforcer<LiftEnforcer>(Binary, Lifted));
```